### PR TITLE
chore(core): allow the bundler to re-use pre-existing bundler output

### LINKED ISSUE #8882: [core] make the asset bundler able to re-use assets
I'd like to request that the asset bundler re-use assets that it has already created if it's instructed to use a source asset hash type.

### Use Case

As it stands, the bundler runs the bundling containers before it checks to see if the assets have already been emitted. This is a problem for me when assets are slow to bundle and when there are several stacks (say, one per deployment environment) in a single app. Having the bundler skip already-emitted assets could save a lot of build and test time. It would also tighten the dev loop, which is slowed by unit tests that drive any of the custom bundling assets that invoke Docker.

### Proposed Solution

I'd suggest that the bundler containers not be re-run when the source asset hash matches an already-existing asset in cdk.out.

### Other

Here's an example project that would benefit from reusing the bundled asset. The stacks run the bundler at least three times to generate the same lambda layer. https://github.com/misterjoshua/aws-cdk-bundle-reuse ([And a github workflow](https://github.com/misterjoshua/aws-cdk-bundle-reuse/runs/832542827?check_suite_focus=true))

I've defined this app:
```ts
const app = new cdk.App();

new BundlerStack(app, 'BundlerStack-Test');
new BundlerStack(app, 'BundlerStack-Stage');
new BundlerStack(app, 'BundlerStack-Production');
```

bundler-stack.ts:
```ts
export class BundlerStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const pythonDir = path.join(__dirname, 'python');

    const libLayer = new lambda.LayerVersion(this, 'LibLayer', {
      code: lambda.Code.fromAsset(pythonDir, {
        exclude: [ '*', '!requirements.txt' ],
        bundling: {
          image: lambda.Runtime.PYTHON_3_8.bundlingDockerImage,
          command: [ 'bash', '-c', `pip install -r requirements.txt -t /asset-output/python`],
        }
      })
    });

    new lambda.Function(this, 'FunOne', {
      code: lambda.Code.fromAsset(pythonDir, { exclude: [ '.venv' ] }),
      handler: 'lambda_one.handler',
      runtime: lambda.Runtime.PYTHON_3_8,
      layers: [ libLayer ],
    });

    new lambda.Function(this, 'FunTwo', {
      code: lambda.Code.fromAsset(pythonDir, { exclude: [ '.venv' ] }),
      handler: 'lambda_two.handler',
      runtime: lambda.Runtime.PYTHON_3_8,
      layers: [ libLayer ],
    });
  }
}
```

`cdk list` takes over a minute for me in this example. The bundler is running three separate pip install containers (one for each stack) even though the results should be identical, based on the source hash. I believe this long time to show a `cdk list` happens because the bundling occurs during the construction phase, rather than during synthesis.

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by misterjoshua**
This PR changes `AssetStaging` so that the bundler will re-use pre-existing output. Before, the bundler would re-run Docker without considering pre-existing assets, which was slow. Now, when handling a `SOURCE` hash type, the bundler detects and returns pre-existing asset output without re-running Docker. For all other hash types, the bundler outputs to a temp directory before calculating asset hashes.

Closes #8882 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/core/test/test.staging.ts [hunk 1]
```diff
@@ -273,6 +342,7 @@ export = {
     // THEN
     test.throws(() => new AssetStaging(stack, 'Asset', {
       sourcePath: directory,
+      assetHashType: AssetHashType.BUNDLE,
```
[the remark this answers is not in the document]
Comment (misterjoshua) [reply packages/@aws-cdk/core/test/test.staging.ts:345]: Ah! Nice catch. I can undo this change. While I was working, this test was failing because the bundle asset already existed in the .cdk.staging from the previous tests, so docker wasn't being run at all so that it could throw - putting it to BUNDLE asset hash type, made it to run every time because it reverted the behaviour to randomly named directories. Now that .cdk.staging is cleaned on test teardown, it's a non-issue. I'll put it back. :)

**Comment by misterjoshua** [comment]
@jogold I was thinking about the failure modes after looking closely at the `throws when bundling fails` test from your review. I think we need to remove the bundleDir if the bundler fails. Otherwise, subsequent runs may falsely believe that the bundleDir can be re-used. I'll push up a commit that does as much in a few minutes.

**Comment by jogold** [comment]
One more thing that comes to mind: what if the `BundlingOptions` are updated? I would expect the asset not be re-used if I change the `command` for example.

Shouldn't we hash those options together with the custom or source hash and use it this final hash in the `bundleDir` name?

cc @eladb

**Comment by eladb** [comment]
> Shouldn't we hash those options together with the custom or source hash and use this final hash in the bundleDir name?

Makes sense

File: packages/@aws-cdk/core/lib/asset-staging.ts [hunk 2]
```diff
@@ -89,10 +89,15 @@ export class AssetStaging extends Construct {
     this.fingerprintOptions = props;
 
     if (props.bundling) {
-      this.bundleDir = this.bundle(props.bundling);
-    }
+      const assetHashIfKnownInAdvance = props.assetHashType !== AssetHashType.BUNDLE
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/asset-staging.ts:92]: please add some comments explaining what this is about

File: packages/@aws-cdk/core/lib/asset-staging.ts [hunk 3]
```diff
@@ -137,15 +142,31 @@ export class AssetStaging extends Construct {
     }
   }
 
-  private bundle(options: BundlingOptions): string {
+  private bundle(options: BundlingOptions, assetHash?: string): string {
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/asset-staging.ts:145]: add a docstring with "@property" explaining why `assetHash` is passed here and what happens if it does not

File: packages/@aws-cdk/core/lib/asset-staging.ts [hunk 4]
```diff
@@ -179,6 +200,7 @@ export class AssetStaging extends Construct {
         workingDirectory: options.workingDirectory ?? AssetStaging.BUNDLING_INPUT_DIR,
       });
     } catch (err) {
+      fs.removeSync(bundleDir);
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/asset-staging.ts:203]: This is unrelated to this PR, right? 
I am also not sure we actually want to delete the bundleDir for diagnosability (I would even include it in the error message).

File: packages/@aws-cdk/core/lib/asset-staging.ts [hunk 5]
```diff
@@ -89,10 +89,15 @@ export class AssetStaging extends Construct {
     this.fingerprintOptions = props;
 
     if (props.bundling) {
-      this.bundleDir = this.bundle(props.bundling);
-    }
+      const assetHashIfKnownInAdvance = props.assetHashType !== AssetHashType.BUNDLE
+        ? this.calculateHash(props)
+        : undefined;
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/asset-staging.ts:94]:
```suggestion
      const sourceHash = props.assetHashType === AssetHashType.SOURCE
        ? this.calculateHash(props)
        : undefined;
```

File: packages/@aws-cdk/core/lib/asset-staging.ts (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/core/lib/asset-staging.ts:145]: Rename to `sourceHash`

File: packages/@aws-cdk/core/lib/asset-staging.ts [hunk 6]
```diff
@@ -89,10 +89,15 @@ export class AssetStaging extends Construct {
...
 
-    this.assetHash = this.calculateHash(props);
+      this.bundleDir = this.bundle(props.bundling, assetHashIfKnownInAdvance);
+      this.assetHash = assetHashIfKnownInAdvance ?? this.calculateHash(props);
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/asset-staging.ts:97]: Not sure I like the fact the calculateHash accepts `props` can we just pass the relevant once for readability?

File: packages/@aws-cdk/core/lib/asset-staging.ts (hunk 4 above)
Comment (jogold) [reply packages/@aws-cdk/core/lib/asset-staging.ts:203]: It is, see https://github.com/aws/aws-cdk/pull/8916#issuecomment-655004255

> Otherwise, subsequent runs may falsely believe that the `bundleDir` can be re-used.

Rename it in case of failure instead (e.g. append `-error`)?

Comment (misterjoshua) [reply packages/@aws-cdk/core/lib/asset-staging.ts:203]: > Rename it in case of failure instead (e.g. append `-error`)?

This seems good to me. But, if we go this route, we might consider removing `-error`  directories if they already exist, such as if docker fails twice for the same bundleDir.

**Comment by misterjoshua** [comment]
@jogold Any idea why the test `throws with assetHash and not CUSTOM hash type` expects Docker to have been invoked?

File: packages/@aws-cdk/core/lib/asset-staging.ts (hunk 6 above)
Comment (misterjoshua) [reply packages/@aws-cdk/core/lib/asset-staging.ts]: I've refactored calculateHash to accept a hashType and custom assetHash parameter which is used when the hashType is CUSTOM.

File: packages/@aws-cdk/core/lib/asset-staging.ts (hunk 3 above)
Comment (misterjoshua) [reply packages/@aws-cdk/core/lib/asset-staging.ts]: Done.

File: packages/@aws-cdk/core/lib/asset-staging.ts (hunk 2 above)
Comment (misterjoshua) [reply packages/@aws-cdk/core/lib/asset-staging.ts]: Done

File: packages/@aws-cdk/core/lib/asset-staging.ts (hunk 4 above)
Comment (misterjoshua) [reply packages/@aws-cdk/core/lib/asset-staging.ts]: For now I've implemented the `-error` suffix. I'm happy to circle back around.

**Comment by jogold** [comment]
> @jogold Any idea why the test `throws with assetHash and not CUSTOM hash type` expects Docker to have been invoked even though it's an error state? What am I missing? Given that the test seems to be to check for a configuration error, the code I've committed assumes this docker run isn't necessary. However, I can circle around to handle this nuance if needed.

It passed because hash calculation always occurred after bundling, it can be safely removed from the expectations. Looks like it was wrongly introduced here https://github.com/aws/aws-cdk/pull/8481.

File: packages/@aws-cdk/core/test/test.staging.ts [hunk 7]
```diff
@@ -133,10 +139,144 @@ export = {
...
+    new AssetStaging(stack, 'AssetWithDifferentBundlingOptions', {
+      sourcePath: directory,
+      bundling: {
+        image: BundlingDockerImage.fromRegistry('alpine'),
+        command: [ DockerStubCommand.SUCCESS ],
+        environment: {
+          UNIQUE_ENV_VAR: 'SOMEVALUE',
+        },
+      },
+    });
```
Comment (jogold) [inline packages/@aws-cdk/core/test/test.staging.ts:212]: Docker runs twice but the logic that we have in `synthesize()` will only move the first asset to the assembly dir (`cdk.out`):
https://github.com/aws/aws-cdk/blob/ba5b268681ccae7c7a948d93494134285e6c112b/packages/%40aws-cdk/core/lib/asset-staging.ts#L116-L126

`targetPath` is based only on the source hash in this case.

Adding an expectation that `cdk.out` should contain both assets would make the test fail here.

Comment (misterjoshua) [reply packages/@aws-cdk/core/test/test.staging.ts:212]: Makes sense. I'll work on this now.

**Comment by misterjoshua** [comment]
> To fix the build error you can run the integ tests in `s3-assets` and `aws-lambda` (the failing tests will be `integ.assets.bundling.lit.js` and `integ.bundling.js` respectively) or merge from master (see #8969).

@jogold I merged from master but the integration tests still failed. I looked at #8969 and added the new pragma to the top of each offending test file to enable the feature. I butchered my commit message, but in any case, those tests seem to work now. I hope that was the right move.

**Comment by eladb** [comment]
Hey @misterjoshua @jogold, we are reverting this PR through #9037 - please revisit.

**Comment by jogold** [comment]
@misterjoshua see https://github.com/aws/aws-cdk/pull/9037#issuecomment-657611401

I suggest waiting for #8962 before revisiting this.

Actually I think it will be easier to implement if the assembly dir is know at construction time.

**Comment by jogold** [comment]
@misterjoshua do you want to give this another try? We can now do everything in `cdk.out`, no need for `.cdk.staging`.

Let me know.

**Comment by misterjoshua** [comment]
> @misterjoshua do you want to give this another try? We can now do everything in `cdk.out`, no need for `.cdk.staging`.
> 
> Let me know.

Hey @jogold. Thanks for letting me know. I'll open a new PR and mention you in it after I've made some progress.
